storage/s3_test.go

```diff
@@ -398,3 +403,200 @@ func TestS3Retry(t *testing.T) {
 		})
 	}
 }
+
+// Credit to aws-sdk-go
+func val(i interface{}, s string) interface{} {
+	v, err := awsutil.ValuesAtPath(i, s)
+	if err != nil || len(v) == 0 {
+		return nil
+	}
+	if _, ok := v[0].(io.Reader); ok {
+		return v[0]
+	}
+
+	if rv := reflect.ValueOf(v[0]); rv.Kind() == reflect.Ptr {
+		return rv.Elem().Interface()
+	}
+
+	return v[0]
+}
+
+// asserts equality; nil interface and empty string are considered equal.
+func assertEqual(t *testing.T, expected string, got interface{}) {
+	if got == nil {
+		if expected != "" {
+			t.Errorf("Expected %q, but received %q", "", got)
+		}
+		return
+	}
+
+	if expected != got {
+		t.Errorf("Expected %q, but received %q", expected, got)
+	}
+}
+
+func TestS3CopyEncryptionRequest(t *testing.T) {
+	testcases := []struct {
+		name     string
+		sse      string
+		sseKeyId string
+
+		expectedSSE      string
+		expectedSSEKeyId string
+		expectedErr      error
+	}{
+		{
+			name: "no encryption, by default",
+		},
+		{
+			name: "aws:kms encryption with server side generated keys",
+			sse:  "aws:kms",
+
+			expectedSSE: "aws:kms",
+		},
+		{
+			name:     "aws:kms encryption with user provided key",
+			sse:      "aws:kms",
+			sseKeyId: "sdkjn12SDdci#@#EFRFERTqW/ke",
+
+			expectedSSE:      "aws:kms",
+			expectedSSEKeyId: "sdkjn12SDdci#@#EFRFERTqW/ke",
+		},
+		{
+			name:     "provide key without encryption flag, shall be ignored",
+			sseKeyId: "1234567890",
+		},
+	}
+	const defaultReqErr = "request was canceled by the user"
```

Comment:
might be defined as type `error` instead of doing `fmt.Errorf` for each case below